### PR TITLE
PERF: avoid copy in _obj_with_exclusions, _selected_obj

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @rhshadrach 

```
import pandas as pd
import numpy as np
import psutil
proc = psutil.Process()

np.random.seed(456123)
df = pd.DataFrame(
    np.random.randn(10_000_000, 5),
    columns=["A", "B", "C", "D", "E"],
)

grps = np.arange(10).repeat(1_000_000)
gb = df.groupby(grps)

gb2 = gb[["C"]]

In [3]: np.shares_memory(gb2._obj_with_exclusions["C"], df["C"])
True   # <- PR
False  # <- main

In [5]: proc.memory_info()
Out[5]: pmem(rss=658894848, vms=36076212224, pfaults=169250, pageins=0)  <- main
Out[5]: pmem(rss=578924544, vms=35993063424, pfaults=149692, pageins=0)  <- PR
```

### CODE REVIEW

File: pandas/core/groupby/groupby.py
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:735]: I think any time `self.obj[self._selection]` results in a Series will be when `self` is a SeriesGroupBy and `self.obj` is a Series, which is handled above.

If the frame has duplicate columns, then we can windup in the case where `self._selection` is hashable and `self` is a DataFrameGroupBy, but in this case I think `self.obj[self_selection]` will then agree with `self._obj_with_exclusions` (both frames).

So I think we can still return obj_with_exclusions here.

**Review by rhshadrach — COMMENTED** [review]
lgtm; I think we can remove a case the logic in _selected_obj, but might be wrong.

File: pandas/core/groupby/groupby.py
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:735]: That logic works for SeriesGroupBy/DataFrameGroupBy, but this code path is shared by Resampler, which defines _gotitem kludgily.  Until we change that, test_resample_groupby_agg_listlike would fail if we return obj_with_exclusions here

Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:735]: > Until we change that

(this is part of why i pinged you about the non-unique columns in agg_dict_like.  If we can make sure to always pass `subset` to `_gotitem`, i think that makes it a lot easier to de-kludge Resample._gotitem)

**Comment by rhshadrach** [comment]
Do you think the memory perf improvement is worth a note in the whatsnew?

**Comment by jbrockmendel** [comment]
> Do you think the memory perf improvement is worth a note in the whatsnew?

added

**Review by rhshadrach — APPROVED** [review]

**Comment by rhshadrach** [comment]
Thanks @jbrockmendel

**Comment by jbrockmendel** [comment]
i think the describe PR will need rebasing (might want to wait until after #51113 is merged) and it should be good to go

**Pull request merged**
